### PR TITLE
 Add AffineConstraints::set_zero for MemorySpace::CUDA 

### PR DESCRIPTION
**Pull request by masterleinad**
We need a separate implementation for AffineConstraints::set_zero in case the data resides on the `CUDA` device.

### CODE REVIEW

File: include/deal.II/lac/affine_constraints.templates.h [hunk 1]
```diff
@@ -1883,6 +1884,49 @@ namespace internal
...
+      std::transform(cm.begin(),
+                     cm.end(),
+                     constrained_local_dofs_host.begin(),
+                     [&vec](const size_type global_index) {
+                       return vec.get_partitioner()->global_to_local(
+                         global_index);
+                     });
+
+      size_type *constrained_local_dofs_device;
+      Utilities::CUDA::malloc(constrained_local_dofs_device, n_constraints);
```
Comment (Rombur) [inline include/deal.II/lac/affine_constraints.templates.h]: You forgot to free the memory.

File: include/deal.II/lac/affine_constraints.templates.h [hunk 2]
```diff
@@ -1883,6 +1884,49 @@ namespace internal
...
+      Utilities::CUDA::copy_to_dev(constrained_local_dofs_host,
+                                   constrained_local_dofs_device);
+
+      const int n_blocks = 1 + (n_constraints - 1) / CUDAWrappers::block_size;
+      set_zero_kernel<<<n_blocks, CUDAWrappers::block_size>>>(
+        constrained_local_dofs_device, cm.size(), vec.get_values());
```
Comment (Rombur) [inline include/deal.II/lac/affine_constraints.templates.h]: You should add an assert that the kernel was launched correctly.

**Comment by masterleinad** [comment]
@Rombur Fixed.

**Pull request merged**